Strip down: EOA transactions, batch counter & validation/execution, 2D nonce manager

forshtat opened the pull request:
These features are important but all together they overload this RIP and make it hard to reason about.
All of these components were moved to RIPs of their own.

forshtat committed:
Strip down: EOA transactions, batch counter & validation/execution, 2D nonce manager

These features are important but all together they overload this RIP
and make it hard to reason about.
All of these components were moved to RIPs of their own.

forshtat committed:
Remove "unused gas penalty" (moved to a separate RIP)

yoavw commented on RIPS/rip-7560.md:
```diff
@@ -811,6 +611,14 @@ the new transaction type.
+when the account executes the `CREATE` (`0xf0`) opcode.
+
+With RIP-7560, the `sender` account's `nonce` will be incremented before a transaction execution starts.
+Developer tooling that rely on counterfactual addresses should take that into account or risk missing the right address.
```
That's only an issue with CREATE though, not CREATE2, right?

forshtat replied:
Right. Added an explanation.

yoavw reviewed (commented):
LGTM, assuming [PR1](https://github.com/eth-infinitism/RIPs/pull/1), [PR3](https://github.com/eth-infinitism/RIPs/pull/3/files) and [PR5](https://github.com/eth-infinitism/RIPs/pull/5/files) include everything that was removed here.

yoavw commented on RIPS/rip-7560.md:
```diff
@@ -836,12 +644,6 @@ All of it creates a lot of new risk vectors and attack surfaces.
-The state that exists at the end of the validation frame may be observed or modified by unrelated contracts before
-the execution frame begins.
-`Sender` contracts must take great care in making sure their code does not make any false assumptions.
-
```
Since we remove validation-execution separation from this RIP, shouldn't we mention the DoS risk on block builders?  Without this separation, this RIP is only safe for chains that don't have a decentralized mempool (such as the single-sequencers used by most rollups), since they look at one transaction at a time and always add it to the current position in the block.  For chains that use a decentralized mempool, cross-transaction invalidations require that separation.

forshtat replied:
Added "Denial of Service attacks on the block builder" security consideration paragraph. I tried to keep it generic and not use any rollup-specific jargon. Please tell me how does it look.

forshtat committed:
Elaborate that CREATE2 addresses do not rely on the account nonce

forshtat committed:
Add "block builder DoS" security consideration

drortirosh commented on RIPS/rip-7560.md:
```diff
@@ -873,6 +673,24 @@ In order to address this developers should track the validity of these transacti
+
+Block builders that rely on a decentralized and permissionless transaction mempool face a risk of having their entire
+transaction mempool being invalidated by a single transaction they pick from it, causing the Denial of Service.
+Sequencer-like block builders do not have such a risk as long as they process incoming transactions one at a time.
```
```suggestion
Sequencer-like block builders do not have such a risk as long as they process incoming transactions in the order they are received.
```

drortirosh commented on RIPS/rip-7560.md:
```diff
@@ -235,21 +138,6 @@ to pay an extra `builderFee` as a "tip" to the block builder.
 This value is denominated in wei and is passed from the `sender`, or the `paymaster` if it is specified,
 to the `coinbase` of the current block as part of the gas pre-charge.
```
this is the first time you mention `coinbase` - it should be applied for general gas payments, above.

L128, above: `callDataCost` can't be compared to `validationGasLimit` (different units)

forshtat replied:
> this is the first time you mention `coinbase` - it should be applied for general gas payments, above.
> 
> L128, above: `callDataCost` can't be compared to `validationGasLimit` (different units)

It seems to be obvious (and defined in EIP-1559) what the meanings of `maxFeePerGas` and `maxPriorityFeePerGas` parameters are. This document does not change that so it doesn't have to redefine these parameters. I will just add a mention that these values are unchanged, I think.

 The `callDataCost` is measured in "gas units" here. Will rename to `callDataGasUsed` to avoid this confusion.

drortirosh commented on RIPS/rip-7560.md:
```diff
@@ -269,13 +156,9 @@ The full list of possible frames tries to replicate the ERC-4337 flow:
 
 All execution frames in the "Validation Phase" must be completed successfully without reverting, and the return value
-for `sender` and `paymaster` validation frames must include `MAGIC_VALUE_SENDER` and `MAGIC_VALUE_PAYMASTER` accrodingly
+for `sender` and `paymaster` validation frames must include `MAGIC_VALUE_SENDER` and `MAGIC_VALUE_PAYMASTER` accordingly
```
the magic return value is not a MUST (as it is part of execution, which we don't control)
Those frames are implicitly reverted if they return anything else.

forshtat replied:
> the magic return value is not a MUST (as it is part of execution, which we don't control) Those frames are implicitly reverted if they return anything else.

We control the block validity. If they return anything else the block is invalid. So I think "MUST" should stay.

drortirosh commented on RIPS/rip-7560.md:
```diff
@@ -284,14 +167,6 @@ In all top-level frames, the global variables have the following meaning:
-
 #### Sender deployment frame
 
 The `deployer` address is invoked with the `deployerData[20:]` as call data input.
```
extract "rationale for separate AA_SENDER_CREATOR" from this, and move into the "rationale" section.

drortirosh replied:
in TransactionType4 type below, missing postOpGasLimit (which we encode into paymasterAndData, along with paymasterGasLimit)

forshtat replied:
Fixed and fixed.

drortirosh commented on RIPS/rip-7560.md:
```diff
@@ -515,64 +390,16 @@ Such transactions MUST NOT be propagated through the default transaction mempool
 and the sending node will be blocked as a spammer.
```
mixing here propagation and execution. should be separated.

forshtat replied:
Removed the paragraph.

drortirosh commented on RIPS/rip-7560.md:
```diff
@@ -515,64 +390,16 @@ Such transactions MUST NOT be propagated through the default transaction mempool
 ### Transaction validity time range parameters
 
 The `Paymaster validation frame` and the `Sender validation frame` each return values `validUntil` and `validAfter`.
-If the transaction is initiated by an EOA, these fields may be encoded into unused bits of the `nonce`.
```
remove this nonce field usage from the sample code too.

forshtat replied:
Fixed.

drortirosh reviewed (commented):
L322: using nonce for range validation? it is not documented anywhere.

RPC methods: should also be extracted.

forshtat committed:
Update RIPS/rip-7560.md

Co-authored-by: Dror Tirosh <dror@opengsn.org>

forshtat committed:
Allign the structure format with the version 0.7 of ERC-4337

1. Split "paymaster"/"deployer" & "paymasterData"/"deployerData" fields
2. Add the explicit "paymasterPostOpGasLimit" field

forshtat committed:
Fix

forshtat committed:
Remove 'block validation' and 'RPC method' parts to separate EIPs

forshtat committed:
Link to the EIP-1559 for the meaning of maxPriorityFeePerGas&maxFeePerGas

forshtat committed:
Rename 'calldataCost' to 'calldataGasUsed' as it is measured in gas units

forshtat committed:
Move the rationale of the "AA_SENDER_CREATOR" to the "Rationale" section

forshtat committed:
Avoid forward-references to the AA transactions' mempool

drortirosh commented on RIPS/rip-7560.md:
```diff
@@ -80,9 +80,12 @@ A new [EIP-2718](https://eips.ethereum.org/EIPS/eip-2718) transaction with type
 0x04 || 0x00 || rlp([
   chainId, sender, nonce, builderFee,
-  callData, paymasterData, deployerData,
+  callData,
```
I try to order the fields in some logical way.
the code flow is  create -> validate -> paymaster

(gas limits below are ordered that way)

forshtat replied:
@drortirosh I changed the order of parameters both here and in the `TransactionType4` solidity struct. Please have a look and tell me if this one makes sense to you. Thanks!

drortirosh commented on RIPS/rip-7560.md:
```diff
@@ -188,9 +187,12 @@ We define the following Solidity struct to represent the AA transaction on-chain
+    uint256 paymasterPostOpGasLimit;
     uint256 callGasLimit;
     uint256 maxFeePerGas;
     uint256 maxPriorityFeePerGas;
```
placing all "bytes" at the end is good, since it allows some optimization when sending the tx.
but still, the ordering is strange.  why paymaster first?

drortirosh commented on RIPS/rip-7560.md:
```diff
@@ -811,6 +445,16 @@ the new transaction type.
+when the account executes the `CREATE` (`0xf0`) opcode.
+
+With RIP-7560, the `sender` account's `nonce` will be incremented before a transaction execution starts.
+Developer tooling that rely on counterfactual addresses should take that into account or risk missing the right address.
```
```suggestion
Developer tooling that rely on counterfactual addresses should take that into account.
```

drortirosh commented on RIPS/rip-7560.md:
```diff
@@ -836,15 +480,10 @@ All of it creates a lot of new risk vectors and attack surfaces.
 ### DoS attacks on block builders
 
-The amount of computation and available memory that is necessary to maintain a mempool and produce valid blocks is
+The amount of computation and available memory that is necessary to produce
```
this section is repeated on L515

forshtat replied:
Fixed.

drortirosh commented on RIPS/rip-7560.md:
```diff
@@ -873,6 +512,25 @@ In order to address this developers should track the validity of these transacti
+
+As part of the mitigation, the AA transactions SHOULD be bound by storage access rules to avoid
+DoS on block builders.
+These rules are defined in [ERC-7562](https://eips.ethereum.org/EIPS/eips/eip-7562).
```
remove full paths: just `(./eip-7562)`, regardless if it is eip/erc/ 
The EIP parser should fix it.

forshtat replied:
The RIP repo is not set up with these tools yet, so it will just be a broken link. I think we will be asked to update the document once the parsers and linters are operational.

yoavw commented on RIPS/rip-7560.md:
```diff
@@ -783,18 +421,14 @@ MUST know which code to execute and understand the calldata provided to them in
 established de-facto standard.
 
-### Accepting `AA_TX_TYPE` transactions from EOAs
+### Calling the `deployer` from the `AA_SENDER_CREATOR` addres
```
```suggestion
### Calling the `deployer` from the `AA_SENDER_CREATOR` address
```

forshtat committed:
Explicitly mention the per-transaction contexts' behaviour

forshtat committed:
Update RIPS/rip-7560.md

Co-authored-by: Dror Tirosh <dror@opengsn.org>

forshtat committed:
Update RIPS/rip-7560.md

Co-authored-by: Yoav Weiss <yoavw@users.noreply.github.com>

forshtat committed:
Change params order; add some security considerations
